File: app/src/main/java/eu/darken/bluemusic/bluetooth/core/BootCheckReceiver.java
package eu.darken.bluemusic.bluetooth.core;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;

import java.util.ArrayList;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.TimeUnit;

import javax.inject.Inject;

import eu.darken.bluemusic.main.core.database.RealmSource;
import eu.darken.bluemusic.settings.core.Settings;
import eu.darken.bluemusic.util.EventGenerator;
import eu.darken.mvpbakery.injection.broadcastreceiver.HasManualBroadcastReceiverInjector;
import io.reactivex.Single;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;

public class BootCheckReceiver extends BroadcastReceiver {

    @Inject Settings settings;
    @Inject BluetoothSource bluetoothSource;
    @Inject EventGenerator eventGenerator;
    @Inject RealmSource realmSource;

    @Override
    public void onReceive(Context context, Intent intent) {
        Timber.v("onReceive(%s, %s)", context, intent);
        if (!Intent.ACTION_BOOT_COMPLETED.equals(intent.getAction())) {
            Timber.e("Triggered with unknown intent: %s", intent);
            return;
        }

        ((HasManualBroadcastReceiverInjector) context.getApplicationContext()).broadcastReceiverInjector().inject(this);

        if (!settings.isEnabled()) {
            Timber.i("We are disabled.");
            return;
        }
        if (!settings.isBootRestoreEnabled()) {
            Timber.i("Restoring on boot is disabled.");
            return;
        }

        Timber.d("We just completed booting, let's see if any Bluetooth device is connected...");
        final PendingResult pendingResult = goAsync();
        bluetoothSource.connectedDevices()
                .subscribeOn(Schedulers.io())
                .map(Map::values)
                .map(ArrayList::new)
                .firstOrError()
                .timeout(8, TimeUnit.SECONDS, Single.just(new ArrayList<>()))
                .doFinally(pendingResult::finish)
                .subscribe(devices -> {
                    Timber.i("Connected devices: %s", devices);

                    final Set<String> managedAddrs = realmSource.getManagedAddresses().blockingGet();
                    boolean hasManagedConnetedDev = false;
                    for (SourceDevice device : devices) {
                        if (managedAddrs.contains(device.getAddress())) {
                            hasManagedConnetedDev = true;
                        }
                    }

                    if (hasManagedConnetedDev) {
                        eventGenerator.send(devices.get(0), SourceDevice.Event.Type.CONNECTED);
                    } else {
                        Timber.d("After boot no managed device is connected.");
                    }

                }, Timber::w);
    }
}

File: app/src/main/java/eu/darken/bluemusic/monitor/core/BootCheckReceiver.kt
package eu.darken.bluemusic.monitor.core

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import dagger.hilt.android.AndroidEntryPoint
import eu.darken.bluemusic.bluetooth.core.BluetoothRepo
import eu.darken.bluemusic.bluetooth.core.currentState
import eu.darken.bluemusic.common.coroutine.DispatcherProvider
import eu.darken.bluemusic.common.datastore.valueBlocking
import eu.darken.bluemusic.common.debug.logging.Logging.Priority.ERROR
import eu.darken.bluemusic.common.debug.logging.Logging.Priority.INFO
import eu.darken.bluemusic.common.debug.logging.Logging.Priority.VERBOSE
import eu.darken.bluemusic.common.debug.logging.asLog
import eu.darken.bluemusic.common.debug.logging.log
import eu.darken.bluemusic.common.debug.logging.logTag
import eu.darken.bluemusic.devices.core.DeviceRepo
import eu.darken.bluemusic.devices.core.DevicesSettings
import eu.darken.bluemusic.monitor.core.service.MonitorControl
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import javax.inject.Inject

@AndroidEntryPoint
class BootCheckReceiver : BroadcastReceiver() {

    @Inject lateinit var devicesSettings: DevicesSettings
    @Inject lateinit var bluetoothSource: BluetoothRepo
    @Inject lateinit var deviceRepo: DeviceRepo
    @Inject lateinit var dispatcherProvider: DispatcherProvider
    @Inject lateinit var monitorControl: MonitorControl

    override fun onReceive(context: Context, intent: Intent) {
        log(TAG, VERBOSE) { "onReceive($context, $intent)" }
        if (Intent.ACTION_BOOT_COMPLETED != intent.action) {
            log(TAG, ERROR) { "Triggered with unknown intent: $intent" }
            return
        }

        if (!devicesSettings.isEnabled.valueBlocking) {
            log(TAG, INFO) { "We are disabled." }
            return
        }
        if (!devicesSettings.restoreOnBoot.valueBlocking) {
            log(TAG, INFO) { "Restoring on boot is disabled." }
            return
        }

        log(TAG) { "We just completed booting, let's see if any Bluetooth device is connected..." }
        val pendingResult = goAsync()

        val scope = CoroutineScope(SupervisorJob() + dispatcherProvider.IO)
        scope.launch {
            try {
                // Wait a bit for Bluetooth to stabilize after boot
                delay(3000)

                val state = bluetoothSource.currentState()

                if (!state.isReady) {
                    log(TAG) { "Bluetooth is not in a ready state: $state" }
                    return@launch
                }

                val connectedDevices = deviceRepo.devices.first().filter { it.isActive }
                log(TAG) { "We booted with already connected devices: $connectedDevices" }

                if (connectedDevices.isEmpty()) {
                    log(TAG, INFO) { "Connected devices are not managed by us." }
                    return@launch
                }

                monitorControl.startMonitor()
            } catch (e: Exception) {
                log(TAG, ERROR) { "Error during boot check: ${e.asLog()}" }
            } finally {
                pendingResult.finish()
                scope.cancel()
            }
        }
    }

    companion object {
        private val TAG = logTag("BootCheckReceiver")
    }
}